bigbang/ingress/abstract.py

```diff
@@ -0,0 +1,773 @@
+            msg["Message-ID"] = archived_at.split("/")[-1]
+        # convert to `EmailMessage` to `mboxMessage`
+        mbox_msg = mboxMessage(msg)
+        mbox_msg.add_header("Archived-At", "<" + archived_at + ">")
+        return mbox_msg
+
+    def from_url(
+        self,
+        list_name: str,
+        url: str,
+        fields: str = "total",
+    ) -> mboxMessage:
+        """
+        Parameters
+        ----------
+        list_name : The name of the mailing list.
+        url : URL of this Email
+        fields : Indicates whether to return 'header', 'body' or 'total'/both or
+            the Email. The latter is the default.
+        """
+        soup = get_website_content(url, session=self.session)
+        if soup == "RequestException":
+            body = "RequestException"
+            header = self.empty_header
+        else:
+            if fields in ["header", "total"]:
+                header = self._get_header_from_html(soup)
+            else:
+                header = self.empty_header
+            if fields in ["body", "total"]:
+                body = self._get_body_from_html(list_name, url, soup)
+            else:
+                body = None
+        return self.create_email_message(url, body, **header)
+
+    @abstractmethod
+    def _get_header_from_html(self, soup: BeautifulSoup) -> Dict[str, str]:
+        pass
+
+    @abstractmethod
+    def _get_body_from_html(
+        self, list_name: str, url: str, soup: BeautifulSoup
+    ) -> Union[str, None]:
+        pass
+
+    @staticmethod
+    def to_dict(msg: mboxMessage) -> Dict[str, List[str]]:
+        """Convert mboxMessage to a Dictionary"""
+        return MessageIO.to_dict(msg)
+
+    @staticmethod
+    def to_pandas_dataframe(msg: mboxMessage) -> pd.DataFrame:
+        """Convert mboxMessage to a pandas.DataFrame"""
+        return MessageIO.to_pandas_dataframe(msg)
+
+    @staticmethod
+    def to_mbox(msg: mboxMessage, filepath: str):
+        """
+        Parameters
+        ----------
+        msg : The Email.
+        filepath : Path to file in which the Email will be stored.
+        """
+        return MessageIO.to_mbox(msg, filepath)
+
+
+class AbstractList(ABC):
+    """
+    This class handles the scraping of a single mailing list of a public archive
+
+    Parameters
+    ----------
+    name : The of whom the list (e.g. 3GPP_COMMON_IMS_XFER, IEEESCO-DIFUSION, ...)
+    source : Contains the information of the location of the mailing list.
+        It can be either an URL where the list or a path to the file(s).
+    msgs : List of mboxMessage objects
+
+    Methods
+    -------
+    from_url()
+    from_messages()
+    from_mbox()
+    get_messages_from_url()
+    get_message_urls()
+    get_period_urls()
+    get_index_of_elements_in_selection()
+    to_dict()
+    to_pandas_dataframe()
+    to_mbox()
+    """
+
+    def __init__(
+        self,
+        name: str,
+        source: Union[List[str], str],
+        msgs: List[mboxMessage],
```

Comment:
Suddenly struck by the problematic ambiguity of the term "List" in light of Python programming type signatures.

Reply:
Would you have another suggestion? The term mailing list is quite common and combined with the SDO of interest, e.g. W3C, the code uses W3CList which might be enough to not be confused with a Python list.

Reply:
I see that this `List` class is from the `typing` library.
Maybe it's unavoidable in this case, and a code comment is therefore appropriate.

See my other comment about how Type Aliases or New Types might be a solution for when data representing the contents of a mailing list is returned as a Python list.